autoprotocol/builders.py

```diff
@@ -1738,6 +1745,154 @@ def blowout(self, volume, initial_z, flowrate=None):
+                    target=Unit(10, "uL/s")
+                ),
+                delay_time=Unit(0.5, "s"),
+                mode_params=LiquidHandle.builders.mode_params(
+                    liquid_class="viscous",
+                    position_z=LiquidHandle.builders.position_z(
+                        reference="preceding_position"
+                    )
+                )
+            )
+        ]
+
+        LiquidHandle.builders.desired_mode(example_transports, None)
+
+        This will return "positive_displacement" based on the "viscous" liquid
+        class. Note that "air" (which defaults to "air_displacement") does not
+        cause a conflict since "air" is often applied in additional transfers
+        of air (such as blowout) for any liquid class, and is disregarded.
+
+        Parameters
+        ----------
+        transports : dict, optional
+            Dictionary of the transport parameters
+        mode : str, optional
+            Mode of dispense type
+
+        Returns
+        -------
+        str
+            Mode of dispense type
+
+        Raises
+        ------
+        TypeError
+            mode is not in str
+        ValueError
+            mode does not contain valid mode value
+        TypeError
+            liquid is not of str
+        ValueError
+            liquid_class is not a valid value
+        ValueError
+            liquid_class did not resolve to a mode
+        ValueError
+            multiple liquid_class exists in one LiquidHandle
+        """
+        # dict for default dispense mode for each liquid_class
+        liquid_class_to_dispense_modes = {
+            "air": "air_displacement",
+            "default": "air_displacement",
+            "viscous": "positive_displacement"
+        }
+        if mode:
+            if not isinstance(mode, str):
+                raise TypeError(
+                    "mode: {} must be in str".format(mode)
+                )
+            if mode not in self.dispense_modes:
+                raise ValueError(
+                    "mode: {} must be one of the valid modes: {}"
+                    "".format(mode, self.dispense_modes)
+                )
+        else:
+            # get liquid_classes from transport input
+            liquid_classes = set([transport["mode_params"]["liquid_class"]
+                                  for transport in transports])
+            # remove automatically added 'air' class from blowout, etc.
+            non_air_classes = [liq for liq in liquid_classes if liq != "air"]
+            # if liquid classes only contain "air" type, mode is returned.
+            if not non_air_classes:
+                return liquid_class_to_dispense_modes["air"]
+
+            modes = []
+            # get mode for each liquid_class if not specified already from the
+            # user input.
+            for liquid in non_air_classes:
+                # resolve mode for different liquid classes (except for None).
+                if liquid:
```

Comment:
Can we filter out the `None`s when populating `non_air_classes` to begin with? Just trying to reduce nesting here, this is still a bit involved.

Reply:
Ya... I contemplated about that, too... Current behavior is that `None` gets `air_displacement` mode by default, so it would not be an issue to filter out both `air` and `None`, but if that behavior changes in the future, then we'll have to worry about having transports with just `air` alone, `None` alone and a mixture of `air` and `None`. I'm passing through all the `None`, and assigning `air_displacement` if the rest of the classes are all `None` explicitly. (`air` can be filtered out at the beginning since `air` can be in the present with any other liquid class types.)

Reply:
Ok, I see, I'm alright with this then.

Reply:
ya... thanks! I think `liquid_handle` just requires some refactoring in general, and when we get around to it, we can improve this. :)